eras/babbage/impl/src/Cardano/Ledger/Babbage/TxInfo.hs

```diff
@@ -97,10 +105,11 @@ txInfoOutV2 txout =
           Right . PV2.OutputDatum . PV2.Datum . PV2.dataToBuiltinData . getPlutusData $ d'
         (SJust _, SJust _) -> Left TranslationLogicErrorDoubleDatum
       referenceScript = transReferenceScript @era $ getField @"referenceScript" txout
-   in case (Alonzo.transTxOutAddr txout, d) of
-        (_, Left e) -> Left e
-        (Nothing, _) -> Left ByronOutputInContext
-        (Just ad, Right d') ->
+   in case (Alonzo.transTxOutAddr txout, d, os) of
+        (_, Left e, _) -> Left e
+        (Nothing, _, OutputFromOutput) -> Left ByronOutputInContext
+        (Nothing, _, OutputFromInput) -> Left ByronInputInContext
+        (Just ad, Right d', _) ->
           Right (PV2.TxOut ad (Alonzo.transValue @(Crypto era) val) d' referenceScript)
```

Comment:
I think the whole thing can be made a little cleaner by switching to `do` notation:

```suggestion
  datum <- 
    case (getField @"datahash" txout, getField @"datum" txout) of
      (SNothing, SNothing) -> Right PV2.NoOutputDatum
      (SJust dh, SNothing) -> Right . PV2.OutputDatumHash $ Alonzo.transDataHash' dh
      (SNothing, SJust d') ->
        Right . PV2.OutputDatum . PV2.Datum . PV2.dataToBuiltinData . getPlutusData $ d'
      (SJust _, SJust _) -> Left TranslationLogicErrorDoubleDatum
   case Alonzo.transTxOutAddr txout of
     Nothing
       | OutputFromOutput <- os -> Left ByronOutputInContext
       | OutputFromInput <- os -> Left ByronInputInContext
     Just ad ->
       Right (PV2.TxOut ad (Alonzo.transValue @(Crypto era) val) datum referenceScript)
```